elasticapm/contrib/starlette/__init__.py

```diff
@@ -108,34 +104,69 @@ class ElasticAPM(BaseHTTPMiddleware):
     >>> elasticapm.capture_message('hello, world!')
     """
 
-    def __init__(self, app: ASGIApp, client: Client):
+    def __init__(self, app: ASGIApp, client: Optional[Client], **kwargs):
         """
 
         Args:
             app (ASGIApp): Starlette app
             client (Client): ElasticAPM Client
         """
-        self.client = client
+        if client:
+            self.client = client
+        else:
+            self.client = make_apm_client(**kwargs)
 
         if self.client.config.instrument and self.client.config.enabled:
             elasticapm.instrumentation.control.instrument()
 
-        super().__init__(app)
-
-    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
-        """Processes the whole request APM capturing.
+        # If we ever make this a general-use ASGI middleware we should use
+        # `asgiref.conpatibility.guarantee_single_callable(app)` here
+        self.app = app
 
+    async def __call__(self, scope, receive, send):
+        """
         Args:
-            request (Request)
-            call_next (RequestResponseEndpoint): Next request process in Starlette.
-
-        Returns:
-            Response
+            scope: ASGI scope dictionary
+            receive: receive awaitable callable
+            send: send awaitable callable
         """
+
+        @functools.wraps(send)
+        async def wrapped_send(message):
+            if message.get("type") == "http.response.start":
+                await set_context(
+                    lambda: get_data_from_response(message, self.client.config, constants.TRANSACTION), "response"
+                )
+                result = "HTTP {}xx".format(message["status"] // 100)
+                elasticapm.set_transaction_result(result, override=False)
+            await send(message)
+
+        # When we consume the body from receive, we replace the streaming
+        # mechanism with a mocked version -- this workaround came from
+        # https://github.com/encode/starlette/issues/495#issuecomment-513138055
+        body = b""
+        while True:
+            message = await receive()
+            if not message:
+                break
+            if message["type"] == "http.request":
+                b = message.get("body", b"")
+                if b:
+                    body += b
+                if not message.get("more_body", False):
+                    break
+            if message["type"] == "http.disconnect":
+                break
```

Comment:
Won't this assemble the whole response in-memory before returning it? That would negate any benefit the user wanted to have by using streaming responses in the first place (e.g. streaming large responses that are built on-the-fly).

Or I misunderstood something, which is quite possible.

Reply:
This is request-side, not response side. And yes, it will construct the whole request body in memory, and then create a wrapped callable to allow the user to also "stream" the request body. This is the only way we can get the request body in middleware -- see https://github.com/encode/starlette/issues/495#issuecomment-513138055

(We were doing this before this PR, it's just uglier now)